README.md
# orka-tools

[![build status](https://github.com/voyages-sncf-technologies/orka-tools/workflows/build/badge.svg)](https://github.com/voyages-sncf-technologies/orka-tools/actions?query=branch%3Amain)

Orka documentation: https://orkadocs.macstadium.com/docs

## Scripts

* `audit_vms.py`: look for "suspicious" VMs that have been running for several hours on an Orka cluster
* `dump_logs.py` & `logs_stats.py`: retrieve & analyse Orka cluster logs
* `orka.py`: an alternate implementation of the Orka CLI that better suits our needs

## Installation

    pip install -r requirements.txt

## Usage

First, you need to define some environment variables:

    export ORKA_CONTROLLER_URL=
    export ORKA_USER_EMAIL=...
    export ORKA_LICENSE_KEY=
    export ORKA_PASSWORD=

You can pass `--help` to any of the scripts to get a detailed description of the arguments & sub-commands it supports.

For example, to quickly connect to a VM through SSH:

    ssh $SSH_USER@$(./orka.py vm get ssh_args --vm $VM_NAME)

Or:

    sshpass -p $SSH_PASSWORD ssh $SSH_USER@$(./orka.py vm get ssh_args --vm $VM_NAME)


CHANGELOG.md
Changelog
---------

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/),
and [PEP 440](https://www.python.org/dev/peps/pep-0440/).


## [?] - not released yet
### Added
- `user list`, `user create` & `user delete` commands in `orka.py`
- `image delete`, `image commit` & `image rename` commands in `orka.py`
- `vm get` command in `orka.py`
- new boolean argument `--force-delete` in `audit_vms.py`
### Removed
- `vm status --vm-only --vm $vm` that became `vm get id --vm $vm` in `orka.py`

## [1.0.1] - 2021-06-16
### Added
* `node list` & `node status` commands in `orka.py`

## [1.0.0] - 2021-06-12
Initial first release


commons.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Orka API reference: https://documenter.getpostman.com/view/6574930/S1ETRGzt

import argparse, os
from contextlib import contextmanager
from getpass import getpass
from urllib.parse import urljoin

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


USER_AGENT = "voyages-sncf-technologies/orka-tools/orka.py"


@contextmanager
def orka_session(orka_controller, user_email, password, license_key, retries=3, backoff_factor=.3, **_):
    'Setup a session with retry adapters, perform login and configure HTTP auth headers'
    session = SessionWithPrefixUrl(orka_controller)
    adapter = HTTPAdapter(max_retries=Retry(total=retries, read=retries, connect=retries, backoff_factor=backoff_factor))
    session.mount('http://', adapter)
    session.mount('https://', adapter)
    resp = check_http_status(session.post('/token', data={
        'email': user_email,
        'password': password,
    }))
    session.headers.update({
        'Authorization': 'Bearer ' + resp.json()['token'],
        'orka-licensekey': license_key,
        'User-Agent': USER_AGENT,
    })
    yield session
    check_http_status(session.delete('/token'))


def check_http_status(response):
    # Very useful to understand errors -> display the HTTP body in case of a non-200 status:
    if response.status_code not in (200, 201):
        print(response.text)
    response.raise_for_status()
    return response


class SessionWithPrefixUrl(requests.Session):
    # Recipe from: https://github.com/psf/requests/issues/2554#issuecomment-109341010
    def __init__(self, prefix_url):
        self.prefix_url = prefix_url
        super().__init__()

    def request(self, method, url, *args, **kwargs):
        url = urljoin(self.prefix_url, url)
        return super().request(method, url, *args, **kwargs)


def add_common_opts_and_parse_args(parser, argv=None):
    parser.add_argument('--orka-controller', help='Default to $ORKA_CONTROLLER_URL')
    parser.add_argument('--license-key', help='Default to $ORKA_LICENSE_KEY')
    parser.add_argument('--user-email', help='Default to $ORKA_USER_EMAIL')
    args = parser.parse_args(argv)
    if not args.orka_controller:
        args.orka_controller = os.environ.get('ORKA_CONTROLLER_URL')
        if not args.orka_controller:
            parser.error('Neither --orka-controller provided nor $ORKA_CONTROLLER_URL set: aborting')
    if not args.license_key:
        args.license_key = os.environ.get('ORKA_LICENSE_KEY')
        if not args.license_key:
            parser.error('Neither --license-key provided nor $ORKA_LICENSE_KEY set: aborting')
    if not args.user_email:
        args.user_email = os.environ.get('ORKA_USER_EMAIL')
        if not args.user_email:
            parser.error('Neither --user-email provided nor $ORKA_USER_EMAIL set: aborting')
    args.password = os.environ.get('ORKA_PASSWORD') or getpass('$ORKA_PASSWORD not found, please provide it: ')
    return args


class ArgparseHelpFormatter(argparse.RawTextHelpFormatter, argparse.ArgumentDefaultsHelpFormatter):
    pass


audit_vms.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
Look for VMs that have been running for several hours on an Orka cluster,
which is suspicious when used as Gitlab CI runners.

Additionally, it can delete those "ghost" VMs.
'''

# USAGE example: ./audit_vms.py --list-running-for-hours 6
# Orka API reference: https://documenter.getpostman.com/view/6574930/S1ETRGzt

import argparse, math, sys
from datetime import datetime

from commons import add_common_opts_and_parse_args, check_http_status, orka_session, ArgparseHelpFormatter


def main(argv):
    args = parse_args(argv)
    ghost_vms_ids = []
    print(" All the times are in UTC timezone")
    print("//-------------------------//")
    present = datetime.utcnow()
    with orka_session(**vars(args)) as session:
        resp = check_http_status(session.get('/resources/vm/list/all'))
        for vm in resp.json()['virtual_machine_resources']:
            if vm['vm_deployment_status'] == 'Not Deployed':
                print(vm["virtual_machine_name"].ljust(22), ':', vm['vm_deployment_status'].ljust(14), " | owner : ", vm['owner'])
                continue
            print(vm["virtual_machine_name"].ljust(22), ':', vm['vm_deployment_status'].ljust(14), " | owner : ", vm['status'][0]['owner'])
            for status in vm['status']:
                print(f'\t{status["virtual_machine_id"]} │ {status["node_location"]} │ {status["virtual_machine_ip"]} │ cpu={status["cpu"]}/{status["vcpu"]} │ {status["RAM"]} │ {status["vm_status"]} │ {status["creation_timestamp"]} │ {status["base_image"]}')
                vm_creation_date = datetime.strptime(status['creation_timestamp'], '%Y-%m-%dT%H:%M:%SZ')
                vm_uptime_in_days = (present - vm_creation_date).days
                vm_uptime_in_hours = math.floor((present - vm_creation_date).seconds / 3600)
                vm_total_uptime_in_hours = (vm_uptime_in_days*24)+vm_uptime_in_hours
                if args.list_running_for_hours and vm_total_uptime_in_hours >= args.list_running_for_hours:
                    ghost_vms_ids.append((status['virtual_machine_id'], vm_uptime_in_days, vm_uptime_in_hours))
            print()
        print()
        if ghost_vms_ids:
            print(f'"ghost" VMs detected, running for at least {args.list_running_for_hours} hours:')
            print('\n'.join(f'{vm_id} - uptime: {uptime_in_days}d and {uptime_in_hours:.1f}h' for vm_id, uptime_in_days, uptime_in_hours in ghost_vms_ids))
            if args.delete_ghost_vms:
                print('You are about to delete all those VMs.')
                if not (args.force_delete or ask_for_confirmation()):
                    print('Aborting')
                    return
                for ghost_vm_id, _, _ in ghost_vms_ids:
                    print('Deleting:', ghost_vm_id)
                    resp = check_http_status(session.delete('/resources/vm/delete',
                                             json={"orka_vm_name": ghost_vm_id}))
                    print(resp.json())
            else:
                sys.exit(2)


def ask_for_confirmation():
    print('Please confirm (y/n): ', end='')
    while True:
        choice = input().lower()
        if choice in ('yes', 'y'):
            return True
        if choice in ('no', 'n'):
            return False
        print("Please respond with 'yes' or 'no': ", end='')


def parse_args(argv=None):
    parser = argparse.ArgumentParser(formatter_class=ArgparseHelpFormatter,
                                     description=__doc__, allow_abbrev=False)
    parser.add_argument('--list-running-for-hours', type=float, help='List VMs running for at least X hours')
    parser.add_argument('--delete-ghost-vms', action='store_true', help='Require --list-running-for-hours')
    parser.add_argument('--force-delete', default=False, action='store_true', help='Bypass interactive confirmation')
    args = add_common_opts_and_parse_args(parser, argv)
    if args.delete_ghost_vms and not args.list_running_for_hours:
        parser.error('--delete-ghost-vms require --list-running-for-hours')
    return args


if __name__ == '__main__':
    main(sys.argv[1:])


dump_logs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
Retrieve all Orka logs from a cluster and dump them into a large JSON file.
It takes a few minutes to complete.
'''

# USAGE: time ./dump_logs.py
# Orka API reference: https://documenter.getpostman.com/view/6574930/S1ETRGzt

import argparse, json, sys

from commons import add_common_opts_and_parse_args, check_http_status, orka_session


def main(argv):
    args = parse_args(argv)
    with orka_session(**vars(args)) as session:
        resp = check_http_status(session.post('/logs/query'))
        with open(args.out_file, 'w', encoding='utf-8') as logs_file:
            json.dump(resp.json(), logs_file, indent=4)


def parse_args(argv=None):
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                                     description=__doc__, allow_abbrev=False)
    parser.add_argument('--out-file', default='logs.json', help=' ')
    return add_common_opts_and_parse_args(parser, argv)


if __name__ == '__main__':
    main(sys.argv[1:])


logs_stats.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'Script producing some statistics on logs frequency, based on a logs.json file extraced with dump_logs.py'

# USAGE example:
# $ ./dump_logs.py && ./logs_stats.py --since 2021-05-20
#     #logs: 119894
#     "Requested CPU is not available in the cluster":
#     runner-xcode-12-5-0 27
#     "Successfully deployed VM":
#     runner-xcode-12-5-0 107
#     runner-xcode-12-4-0 18
#     "Successfully deleted VM(s)":
#     runner-xcode-12-5-0 107
#     runner-xcode-12-4-0 15

import argparse, json, sys
import re
from collections import defaultdict
from datetime import datetime


TARGET_MESSAGES = (
    'Requested CPU is not available in the cluster',
    'Successfully deployed VM',
    'Successfully deleted VM(s)',
)


def main(argv=None):
    args = parse_args(argv)

    with open(args.logs_filename, encoding='utf-8') as logs_file:
        logs = json.load(logs_file)['logs']
    print('#logs:', len(logs))

    if args.for_vm:
        for log in logs:
            log_date = datetime.strptime(log['createdAt'][:-5], '%Y-%m-%dT%H:%M:%S')
            if args.since and log_date < args.since:
                continue
            if log['request']['body'].get('orka_vm_name') == args.for_vm:
                print(json.dumps(log, indent=4))
        return

    for vm_deploy_msg in TARGET_MESSAGES:
        vms = defaultdict(int)
        for log in logs:
            log_date = datetime.strptime(log['createdAt'][:-5], '%Y-%m-%dT%H:%M:%S')
            if args.since and log_date < args.since:
                continue
            body = log['response']['body']
            if isinstance(body, dict):
                ok_msg_match = body.get('message') == vm_deploy_msg
                error_msg_match = any(re.match(vm_deploy_msg, error['message']) for error in body.get('errors', []))
                if ok_msg_match or error_msg_match:
                    orka_vm_name = body.get('help', {}).get('required_request_data_for_deploy', {}).get('orka_vm_name') or log['request']['body']['orka_vm_name']
                    vms[orka_vm_name] += 1
        print(f'"{vm_deploy_msg}":')
        for vm_name, count in sorted(vms.items(), key=lambda vm: vm[1], reverse=True):
            print(vm_name, count)


def parse_args(argv=None):
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                                     description=__doc__, allow_abbrev=False)
    parser.add_argument('--for-vm', help='Display all logs matching this VM name & exit')
    parser.add_argument('--logs-filename', default='logs.json', help=' ')
    parser.add_argument('--since', help='Date must be specified with this format: YYYY-MM-DD')
    args = parser.parse_args(argv)
    if args.since:
        args.since = datetime.strptime(args.since, '%Y-%m-%d')
    return args


if __name__ == '__main__':
    main(sys.argv[1:])


orka.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
This script is an alternate implementation of the Orka CLI:
https://orkadocs.macstadium.com/docs/quick-start

It alleviates the following limitations of the official Orka CLI:
* the official Orka CLI always return a ZERO error code;
  this CLI return a non-ZERO error code in case of failure.
* a large majority of the official Orka CLI require a TTY to be used correctly,
  making it difficult to use in scripts; this CLI does not require a TTY.
* the official Orka CLI produces ASCII coloring control sequences,
  making its output difficult to parse; this CLI only produces plain text.

All the sub-commands of this CLI are similar to the ones of the official Orka CLI.
However, this CLI it does not implement ALL sub-commands of the official Orka CLI so far,
as we only required a subset of them in our Ansible playbook.
'''

# USAGE: ./orka.py --help
# Orka API reference: https://documenter.getpostman.com/view/6574930/S1ETRGzt

import argparse, json, sys
from getpass import getpass

from commons import add_common_opts_and_parse_args, check_http_status, orka_session, ArgparseHelpFormatter


def main(argv):
    args = parse_args(argv)
    with orka_session(**vars(args)) as session:
        args.func(args, session)


def vm_list(_, session):
    resp = check_http_status(session.get('/resources/vm/list/all'))
    deployed_vms = [vm for vm in resp.json()["virtual_machine_resources"] if "status" in vm]
    print(f"Deployed: {len(deployed_vms)}")
    if deployed_vms:
        print('        Name        │      ID       │       Owner       │     Node      │      IP       │ VNC  │ SSH  │ vCPU/CPU │ RAM │          Base image           │ Status  │ Deploy Date')
    for vm in deployed_vms:
        status = vm["status"][0]
        print(f"{vm['virtual_machine_name']:<19} | {status['virtual_machine_id']:<13} | {status['owner']:<17} | {status['node_location']:<8} ({status['node_status']:<2}) | {status['virtual_machine_ip']:<13} | {status['vnc_port']} | {status['ssh_port']} | {status['cpu']}/{status['vcpu']}      | {status['RAM']:<3} | {status['base_image']:<29} | {status['vm_status']:<7} | {status['creation_timestamp']}")
    print()
    not_deployed_vms = [vm for vm in resp.json()["virtual_machine_resources"] if "status" not in vm]
    print(f"Not Deployed: {len(not_deployed_vms)}")
    if not_deployed_vms:
        print('        Name        │       Owner       │ vCPU/CPU │ Base image')
    for vm in not_deployed_vms:
        print(f"{vm['virtual_machine_name']:<19} | {vm['owner']:<17} | {vm['cpu']}/{vm['vcpu']}      | {vm['base_image']}")

def vm_status(args, session):
    resp = check_http_status(session.get(f'/resources/vm/status/{args.vm}'))
    vms = resp.json()["virtual_machine_resources"]
    if not vms:
        print("No VMs found matching this name")
        sys.exit(1)
    print(json.dumps(vms, indent=4))

def vm_get(args, session):
    resp = check_http_status(session.get(f'/resources/vm/status/{args.vm}'))
    vms = resp.json()["virtual_machine_resources"]
    if not vms:
        print("No VMs found matching this name")
        sys.exit(1)
    if len(vms) > 1:
        print("Several VMs match this name")
        sys.exit(1)
    vm = vms[0]
    if "status" not in vm:
        raise RuntimeError(f"VM {args.vm} not deployed")
    status = vm["status"][0]
    if args.field == "id":
        print(status["virtual_machine_id"])
    elif args.field == "ssh_args":
        print(status["virtual_machine_ip"] + " -p " + status["ssh_port"])

def vm_create_config(args, session):
    resp = check_http_status(session.post('/resources/vm/create',
                                          json={
                                            "orka_base_image": args.base_image,
                                            "orka_vm_name": args.vm,
                                            "orka_image": args.vm,
                                            "orka_cpu_core": args.cpu,
                                            "vcpu_count": args.vcpu,
                                          }))
    print(json.dumps(resp.json(), indent=4))

def vm_deploy(args, session):
    resp = check_http_status(session.post('/resources/vm/deploy',
                                          json={"orka_vm_name": args.vm}))
    print(json.dumps(resp.json(), indent=4))

def vm_create(args, session):
    vm_create_config(args, session)
    vm_deploy(args, session)

def vm_suspend(args, session):
    resp = check_http_status(session.post('/resources/vm/exec/suspend',
                                          json={"orka_vm_name": args.vm}))
    print(json.dumps(resp.json(), indent=4))

def vm_delete(args, session):
    resp = check_http_status(session.delete('/resources/vm/delete',
                                            json={"orka_vm_name": args.vm}))
    print(json.dumps(resp.json(), indent=4))

def vm_purge(args, session):
    resp = check_http_status(session.delete('/resources/vm/purge',
                                            json={"orka_vm_name": args.vm}))
    print(json.dumps(resp.json(), indent=4))


def image_list(_, session):
    resp = check_http_status(session.get('/resources/image/list'))
    images = resp.json()["image_attributes"]
    if images:
        print("             image             │ image_size │         modified         │        date_added        │ owner")
    for img in images:
        print(f"{img['image']:<30} | {img['image_size']:<3}        │ {img['modified']:<24} | {img['date_added']:<24} | {img['owner']}")

def image_save(args, session):
    resp = check_http_status(session.post('/resources/image/save',
                                          json={
                                            "orka_vm_name": args.vm_id,
                                            "new_name": args.new_base_image_name
                                          }))
    print(json.dumps(resp.json(), indent=4))

def image_commit(args, session):
    resp = check_http_status(session.post('/resources/image/commit',
                                          json={"orka_vm_name": args.vm_id}))
    print(json.dumps(resp.json(), indent=4))

def image_rename(args, session):
    resp = check_http_status(session.post('/resources/image/rename',
                                          json={
                                            "image": args.image,
                                            "new_name": args.o
                                          }))
    print(json.dumps(resp.json(), indent=4))

def image_delete(args, session):
    resp = check_http_status(session.post('/resources/image/delete', json={"image": args.image}))
    print(json.dumps(resp.json(), indent=4))


def node_list(_, session):
    resp = check_http_status(session.get('/resources/node/list'))
    nodes = resp.json()["nodes"]
    if nodes:
        print("   Node    │      IP       │  CPU  │ Memory │ State")
    for node in nodes:
        print(f"{node['name']:<10} | {node['address']} │ {node['available_cpu']:>2}/{node['allocatable_cpu']} | {node['available_memory']:<6} | {node['state']}")

def node_status(args, session):
    resp = check_http_status(session.get(f'/resources/node/status/{args.node}'))
    print(json.dumps(resp.json(), indent=4))


def user_list(_, session):
    resp = check_http_status(session.get('/users'))
    for group, users in resp.json()["user_groups"].items():
        print(group)
        for user in users:
            print(f'  {user}')

def user_create(args, session):
    print('Password: ', end='')
    payload = {"email": args.email, "password": getpass()}
    if args.group:
        payload['group'] = args.group
    resp = check_http_status(session.post('/users/', json=payload))
    print(json.dumps(resp.json(), indent=4))

def user_delete(args, session):
    resp = check_http_status(session.delete(f'/users/{args.email}'))
    print(json.dumps(resp.json(), indent=4))


def parse_args(argv=None):
    # pylint: disable=too-many-statements
    parser = argparse.ArgumentParser(formatter_class=ArgparseHelpFormatter,
                                     description=__doc__, allow_abbrev=False)
    parser.add_argument('--retries', type=int, default=3, help='Max HTTP retries')
    parser.add_argument('--backoff-factor', type=float, default=.3, help='Backup factor for HTTP retries')
    subparsers = parser.add_subparsers(required=True)

    vm_cmd = subparsers.add_parser('vm')
    vm_subparsers = vm_cmd.add_subparsers(dest='vm', required=True)
    vm_list_cmd = vm_subparsers.add_parser('list')
    vm_list_cmd.set_defaults(func=vm_list)
    vm_status_cmd = vm_subparsers.add_parser('status')
    vm_status_cmd.set_defaults(func=vm_status)
    vm_status_cmd.add_argument('--vm', '-v', required=True)
    vm_get_cmd = vm_subparsers.add_parser('get')
    vm_get_cmd.set_defaults(func=vm_get)
    vm_get_cmd.add_argument('field', choices=('id', 'ssh_args'))
    vm_get_cmd.add_argument('--vm', '-v', required=True)
    vm_create_config_cmd = vm_subparsers.add_parser('create-config')
    vm_create_config_cmd.set_defaults(func=vm_create_config)
    vm_create_config_cmd.add_argument('--vm', '-v', required=True)
    vm_create_config_cmd.add_argument('--base-image', '-b', required=True)
    vm_create_config_cmd.add_argument('--cpu', '-c', type=int, required=True)
    vm_create_config_cmd.add_argument('--vcpu', '-C', type=int, required=True)
    vm_deploy_cmd = vm_subparsers.add_parser('deploy')
    vm_deploy_cmd.set_defaults(func=vm_deploy)
    vm_deploy_cmd.add_argument('--vm', '-v', required=True)
    vm_create_cmd = vm_subparsers.add_parser('create')
    vm_create_cmd.set_defaults(func=vm_create)
    vm_create_cmd.add_argument('--vm', '-v', required=True)
    vm_create_cmd.add_argument('--base-image', '-b', required=True)
    vm_create_cmd.add_argument('--cpu', '-c', type=int, required=True)
    vm_create_cmd.add_argument('--vcpu', '-C', type=int, required=True)
    vm_suspend_cmd = vm_subparsers.add_parser('suspend')
    vm_suspend_cmd.set_defaults(func=vm_suspend)
    vm_suspend_cmd.add_argument('--vm', '-v', required=True)
    vm_delete_cmd = vm_subparsers.add_parser('delete')
    vm_delete_cmd.set_defaults(func=vm_delete)
    vm_delete_cmd.add_argument('--vm', '-v', required=True)
    vm_purge_cmd = vm_subparsers.add_parser('purge')
    vm_purge_cmd.set_defaults(func=vm_purge)
    vm_purge_cmd.add_argument('--vm', '-v', required=True)

    img_cmd = subparsers.add_parser('image')
    img_subparsers = img_cmd.add_subparsers(dest='image', required=True)
    img_list_cmd = img_subparsers.add_parser('list')
    img_list_cmd.set_defaults(func=image_list)
    img_save_cmd = img_subparsers.add_parser('save')
    img_save_cmd.set_defaults(func=image_save)
    img_save_cmd.add_argument('--vm-id', '--vm', '-v', required=True)
    img_save_cmd.add_argument('--new-base-image-name', '-b', required=True)
    img_delete_cmd = img_subparsers.add_parser('delete')
    img_delete_cmd.set_defaults(func=image_delete)
    img_delete_cmd.add_argument('-i', '--image', required=True)
    img_commit_cmd = img_subparsers.add_parser('commit')
    img_commit_cmd.set_defaults(func=image_commit)
    img_commit_cmd.add_argument('--vm-id', '--vm', '-v', required=True)
    img_rename_cmd = img_subparsers.add_parser('rename')
    img_rename_cmd.set_defaults(func=image_rename)
    img_rename_cmd.add_argument('-i', '--image', required=True)
    img_rename_cmd.add_argument('-o', required=True)

    node_cmd = subparsers.add_parser('node')
    node_subparsers = node_cmd.add_subparsers(dest='node', required=True)
    node_list_cmd = node_subparsers.add_parser('list')
    node_list_cmd.set_defaults(func=node_list)
    node_status_cmd = node_subparsers.add_parser('status')
    node_status_cmd.set_defaults(func=node_status)
    node_status_cmd.add_argument('--node', '-n', required=True)

    user_cmd = subparsers.add_parser('user')
    user_subparsers = user_cmd.add_subparsers(dest='user', required=True)
    user_list_cmd = user_subparsers.add_parser('list')
    user_list_cmd.set_defaults(func=user_list)
    user_create_cmd = user_subparsers.add_parser('create')
    user_create_cmd.add_argument('-e', dest='email', required=True)
    user_create_cmd.add_argument('-g', dest='group')
    user_create_cmd.set_defaults(func=user_create)
    user_delete_cmd = user_subparsers.add_parser('delete')
    user_delete_cmd.add_argument('-e', dest='email', required=True)
    user_delete_cmd.set_defaults(func=user_delete)

    return add_common_opts_and_parse_args(parser, argv)


if __name__ == '__main__':
    main(sys.argv[1:])
